File: kvstore-3/py/kvstore/wal.py
"""Write-Ahead Log implementation for persistence."""

import json
import os
from typing import Any, Optional
from .utils import get_current_timestamp


class WAL:
    """Write-Ahead Log for persisting operations.

    Logs all write operations (set, delete) to an append-only file.
    Supports replay on startup for crash recovery.
    """

    def __init__(self, data_dir: str):
        """Initialize the WAL.

        Args:
            data_dir: Directory where the WAL file will be stored
        """
        self.data_dir = data_dir
        self.wal_path = os.path.join(data_dir, 'wal.log')

        os.makedirs(data_dir, exist_ok=True)

        if not os.path.exists(self.wal_path):
            open(self.wal_path, 'a').close()

    def log_set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Log a set operation.

        Args:
            key: The key being set
            value: The value being stored
            ttl: Optional TTL in seconds
        """
        entry = {
            'op': 'set',
            'key': key,
            'value': value,
            'timestamp': get_current_timestamp()
        }
        if ttl is not None:
            entry['ttl'] = ttl

        with open(self.wal_path, 'a') as f:
            f.write(json.dumps(entry) + '\n')

    def log_delete(self, key: str) -> None:
        """Log a delete operation.

        Args:
            key: The key being deleted
        """
        entry = {
            'op': 'delete',
            'key': key,
            'timestamp': get_current_timestamp()
        }

        with open(self.wal_path, 'a') as f:
            f.write(json.dumps(entry) + '\n')

    def replay(self, store) -> None:
        """Replay all operations from the WAL to restore state.

        Args:
            store: KVStore instance to replay operations into
        """
        if not os.path.exists(self.wal_path):
            return

        with open(self.wal_path, 'r') as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue

                try:
                    entry = json.loads(line)
                    op = entry.get('op')

                    if op == 'set':
                        key = entry['key']
                        value = entry['value']
                        ttl = entry.get('ttl')
                        timestamp = entry.get('timestamp', get_current_timestamp())

                        if ttl is not None:
                            expiration_time = timestamp + ttl
                            remaining_ttl = expiration_time - get_current_timestamp()

                            if remaining_ttl > 0:
                                store.set(key, value, ttl=remaining_ttl)
                        else:
                            store.set(key, value)

                    elif op == 'delete':
                        key = entry['key']
                        store.delete(key)

                except (json.JSONDecodeError, KeyError):
                    continue

File: kvstore-4/ts/src/wal.ts
/**
 * Write-Ahead Log (WAL) implementation for persistence
 */

import * as fs from 'fs/promises';
import * as path from 'path';
import { getCurrentTimestamp } from './utils';

export interface WALOperation {
  op: 'set' | 'delete';
  key: string;
  value?: any;
  ttl?: number;
  timestamp: number;
}

export class WAL {
  private filePath: string;
  private fileHandle: fs.FileHandle | null = null;

  constructor(dataDir: string) {
    this.filePath = path.join(dataDir, 'wal.log');
  }

  /**
   * Initialize the WAL file
   */
  async init(): Promise<void> {
    // Ensure the directory exists
    const dir = path.dirname(this.filePath);
    await fs.mkdir(dir, { recursive: true });

    // Open file in append mode
    this.fileHandle = await fs.open(this.filePath, 'a');
  }

  /**
   * Append an operation to the WAL
   */
  async append(op: 'set' | 'delete', key: string, value?: any, ttl?: number): Promise<void> {
    const operation: WALOperation = {
      op,
      key,
      timestamp: getCurrentTimestamp(),
    };

    if (op === 'set') {
      operation.value = value;
      if (ttl !== undefined && ttl !== null) {
        operation.ttl = ttl;
      }
    }

    const line = JSON.stringify(operation) + '\n';

    if (this.fileHandle) {
      await this.fileHandle.write(line);
      // Flush to disk to ensure durability
      await this.fileHandle.datasync();
    }
  }

  /**
   * Replay the WAL and return all operations
   */
  async replay(): Promise<WALOperation[]> {
    try {
      const content = await fs.readFile(this.filePath, 'utf-8');
      const lines = content.trim().split('\n').filter(line => line.length > 0);
      return lines.map(line => JSON.parse(line));
    } catch (err: any) {
      if (err.code === 'ENOENT') {
        // File doesn't exist yet, return empty array
        return [];
      }
      throw err;
    }
  }

  /**
   * Close the WAL file
   */
  async close(): Promise<void> {
    if (this.fileHandle) {
      await this.fileHandle.close();
      this.fileHandle = null;
    }
  }
}
